### PR TITLE
Adds post on Dask's hyperparameter optimization

### LINKED ISSUE #33: Blog post on Dask-ML's hyperparameter optimization
It'd be good to show the work performed on Dask-ML and `HyperbandSearchCV`. It's recently been merged in https://github.com/dask/dask-ml/pull/221.

### PR DESCRIPTION
**Pull request by stsievert**
This PR closes #33. It adds a post on the work I did with Hyperband over the summer.

### CODE REVIEW

**Comment by stsievert** [comment]
Let me know what issues you have with the header/note at the top. Right now it says

>  *Scott Sievert wrote this post. The original post lives at https://stsievert.com/blog/2019/09/19/dask-hyperparam-opt/ with better styling. This work is supported by Ananconda, Inc.*

This is a copy of the line @jcrist used at https://blog.dask.org/2016/07/12/dask-learn-part-1. Should I include "This work is supported by Anaconda, Inc"?

~~That post is live, but haven't published it anywhere (RSS feed or twitter). I plan on keeping it that way until this PR is merged.~~ (edit: I've taken it down).

Also, I don't explicitly mention any of @jcrist's work on dask-searchcv. I do implicitly mention it in future work.

File: _posts/2019-09-19-dask-hyperparam-opt.md [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+model and "hyperparameters", parameters that are chosen before training begins that
+help with cohesion between the model and data. The user needs to specify values
+for these hyperparameters in order to use the model. A good example is
+adapting ridge regression or LASSO to the amount of noise in the
+data with the regularization parameter.[^alpha]
+
+[^alpha]:Which amounts to choosing `alpha` in Scikit-learn's [Ridge](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Ridge.html) or [LASSO](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Lasso.html)
+
+Model performance strongly depends on the hyperparameters provided.  A simple
+example is with tuning the regularization parameter in ridge regression or
```
Comment (TomAugspurger) [inline _posts/2019-09-19-dask-hyperparam-opt.md]: ~with~

File: _posts/2019-09-19-dask-hyperparam-opt.md [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+LASSO.[^regularization] Another more complex example is with a particular
```
Comment (TomAugspurger) [inline _posts/2019-09-19-dask-hyperparam-opt.md]: ~with~

File: _posts/2019-09-19-dask-hyperparam-opt.md [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+advanced task scheduler like Dask that I'm aware of.[^other-work]
+
+[^new]:By some metrics. It's been around since 2016 and some call that "old news."
+[^other-work]:Including with [Ray-Tune's implementation][asha-ray] and "[Massively Parallel Hyperparameter Tuning][asha]".
+[asha-ray]:https://ray.readthedocs.io/en/latest/tune-schedulers.html#asynchronous-hyperband
+[asha]:https://arxiv.org/pdf/1810.05934.pdf
+
+I'll highlight some key points of the corresponding paper,
+"[Better and faster hyperparameter optimization with Dask][scipy19]".  Let's go
+through the basics of Hyperband then illustrate it's use and performance with
```
Comment (TomAugspurger) [inline _posts/2019-09-19-dask-hyperparam-opt.md]: it's -> its.

File: _posts/2019-09-19-dask-hyperparam-opt.md [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+an example.
+
+## Hyperband basics
+
+The motivation for Hyperband is to high performing models with minimal
```
Comment (TomAugspurger) [inline _posts/2019-09-19-dask-hyperparam-opt.md]: Missing a word: "find"?

File: _posts/2019-09-19-dask-hyperparam-opt.md [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+training. Given this goal, it makes sense to spend more time training high
+performing models – why waste training time on poor performing models?
```
Comment (TomAugspurger) [inline _posts/2019-09-19-dask-hyperparam-opt.md]: May be worth introducing the incremental concept here? If we *knew* that it's a poor model, then we wouldn't need hyperparameter opt in the first place. But the intuition is that a model that starts out poor will probably be poor on future batches of data.

File: _posts/2019-09-19-dask-hyperparam-opt.md [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+>>> search = HyperbandSearchCV(
+...     est, params, max_iter=max_iter, aggressiveness=4
+... )
+```
+
+`aggressiveness` defaults to 3. `aggressiveness=4` is chosen because this is an
+*initial* search; I know nothing about how this search space. Then, this search
+should be more aggressive in culling off bad models.
+
+I may be worried about how long the computation will take, or how many models
```
Comment (TomAugspurger) [inline _posts/2019-09-19-dask-hyperparam-opt.md]: May want to explain metadata before showing it.

File: _posts/2019-09-19-dask-hyperparam-opt.md [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+>>> search.fit(X_train, y_train)
+```
+
+The dashboard will be active during this time[^dashboard]:
+
+[^dashboard]:But it probably won't be this fast: the video is sped up by a factor of 3.
+
+<p>
+<video width="600" autoplay loop controls class="center">
+  <source src="/images/2019-hyperband/dashboard-compress.mp4" type="video/mp4" class="center">
```
Comment (TomAugspurger) [inline _posts/2019-09-19-dask-hyperparam-opt.md]: Do we have other videos in this repository? Or can we upload it to youtube and embed the video?

**Review by TomAugspurger — COMMENTED** [review]
Looks nice, thanks.

I wonder if we can cut down the section introducing hyperparameter optimization a bit? Just to get the length down?

File: _posts/2019-09-19-dask-hyperparam-opt.md (hunk 5 above)
Comment (stsievert) [reply _posts/2019-09-19-dask-hyperparam-opt.md]: Rewritten as

> training. Given this goal, it makes sense to spend more time training high performing models – why waste more time training time a model if it's done poorly in up until now?

File: _posts/2019-09-19-dask-hyperparam-opt.md (hunk 7 above)
Comment (stsievert) [reply _posts/2019-09-19-dask-hyperparam-opt.md]: Currently, I only have that one video (which depends on https://github.com/dask/dask-ml/pull/528).

I'm not opposed to making another video.

Comment (TomAugspurger) [reply _posts/2019-09-19-dask-hyperparam-opt.md]: Given that the video is so small, it's probably fine to include it here in the git repostitory.

Comment (stsievert) [reply _posts/2019-09-19-dask-hyperparam-opt.md]: Ah, gotcha. I only include a compressed version in git (I think 700kB). The uncompressed version is too large for git, 27 MB.

**Comment by stsievert** [comment]
Thanks for the comments @TomAugspurger! I've addressed them all, and made the "problem" section shorter (by deleting an unnecessary paragraph).

**Comment by TomAugspurger** [comment]
I think we're pretty good here. Anything else you wanted to change @stsievert?

**Comment by stsievert** [comment]
Yeah, I'd like to add a link explaining what Dask is – this post assumes the user knows what "Dask, an advanced task scheduler is." What pages would be good to link to? I'm thinking of linking to https://docs.dask.org/en/latest/why.html for a general overview.

**Comment by TomAugspurger** [comment]
Yeah, that's probably a reasonable place to link.

**Comment by stsievert** [comment]
Done. I also link to the video introductions to Dask and Dask-ML.

I *think* everything looks good – links go where I expect, images appear, etc. @TomAugspurger could you double-check?

**Comment by stsievert** [comment]
I have a couple small edits to make; give me another hour.

**Comment by stsievert** [comment]
I've made the edits and fixed some obvious typos. This is ready for review now.

**Comment by TomAugspurger** [comment]
I won't have time to look at this again for another couple weeks, but I think it was ready to go.

Can you just update the filename to today's date and we'll get this merged?

**Comment by stsievert** [comment]
> update the filename to today's date

I've updated it to today's date (Sept. 26th) and same with the post on my own site. I can update it tomorrow with tomorrow's date too.

**Comment by stsievert** [comment]
I've updated the name to reflect today's date (Sept. 27th).

**Comment by TomAugspurger** [comment]
Thanks @stsievert!
